app/queries/drug_stocks_query.rb

```diff
@@ -0,0 +1,103 @@
+class DrugStocksQuery
+  CACHE_VERSION = 1
+
+  def initialize(facilities:, for_end_of_month:)
+    @facilities = facilities
+    @for_end_of_month = for_end_of_month
+    # assuming that all facilities on the page have the same protocol
+    @protocol = @facilities.first.protocol
+    @state = @facilities.first.state
+  end
+
+  def protocol_drugs_by_category
+    @protocol_drugs_by_category ||= @protocol.protocol_drugs
+      .where(stock_tracked: true)
+      .order(:name, :dosage)
+      .sort_by(&:sort_key)
+      .group_by(&:drug_category)
+      .sort_by { |(drug_category, _)| drug_category }
+      .to_h
+  end
+
+  def cache_key
+    [
+      self.class.name,
+      @facilities.map(&:id).sort,
+      @for_end_of_month,
+      @protocol.id,
+      @state,
+      CACHE_VERSION
+    ].join("/")
+  end
+
+  def call
+    Rails.cache.fetch(cache_key,
```

Comment:
Have we considered caching at the per-facility level? It's more cache keys, but each cache key is pretty neatly scoped, and may set us up nicely for future iterations like "show me all large facilities", or "show me all facilities in a block", etc.

Reply:
This report is generally not very expensive, since the drug stocks per facility per month aren't going to be that many and the patient days calculation per facility isn't too intense. So we'll probably get very little out of caching it at a facility level. This is also why we didn't think too much about busting the cache anytime a drug stock is added for a facility in the list.

It's still somewhat useful to cache at a block level for example (that's the default view) since there's totalling and there might be a few thousand facilities in some blocks. And because we generally want the page loads to be quick.

Reply:
Cool cool 👍 The cache keys become unwieldy but we don't need worry about that, right? We never interact with them directly.